File: examples/python/m2-tool-runtime/tools/read.py
"""read：读取工作区内文件内容。"""

from __future__ import annotations

from typing import Any, Mapping

from tool_types import ToolSpec
from workspace import Workspace, WorkspaceError

DEFAULT_MAX_LINES = 2000
DEFAULT_MAX_BYTES = 512_000


class ReadTool:
    """按路径读取文本文件，支持 offset/limit 截断。"""

    def __init__(self, workspace: Workspace) -> None:
        self.workspace = workspace
        self.spec = ToolSpec(
            name="read",
            description="Read file contents inside the workspace",
            input_schema={
                "type": "object",
                "required": ["path"],
                "properties": {
                    "path": {"type": "string"},
                    "offset": {"type": "integer"},
                    "limit": {"type": "integer"},
                },
                "additionalProperties": False,
            },
        )

    # ANCHOR: m2-read-tool
    def validate_arguments(self, arguments: Mapping[str, Any]) -> None:
        path = arguments.get("path")
        if not isinstance(path, str) or not path.strip():
            raise ValueError("path must be a non-empty string")
        offset = arguments.get("offset")
        if offset is not None and (not isinstance(offset, int) or isinstance(offset, bool) or offset < 1):
            raise ValueError("offset must be a positive integer")
        limit = arguments.get("limit")
        if limit is not None and (not isinstance(limit, int) or isinstance(limit, bool) or limit < 1):
            raise ValueError("limit must be a positive integer")

    def execute(self, arguments: Mapping[str, Any]) -> dict[str, Any]:
        path = self.workspace.resolve(str(arguments["path"]), must_exist=True)
        if not path.is_file():
            raise WorkspaceError(f"not a file: {arguments['path']!r}")

        data = path.read_bytes()
        truncated_bytes = False
        if len(data) > DEFAULT_MAX_BYTES:
            data = data[:DEFAULT_MAX_BYTES]
            truncated_bytes = True

        text = data.decode("utf-8", errors="replace")
        lines = text.splitlines()
        offset = int(arguments.get("offset") or 1)
        start = max(offset - 1, 0)
        limit = arguments.get("limit")
        max_lines = int(limit) if limit is not None else DEFAULT_MAX_LINES
        selected = lines[start : start + max_lines]
        truncated_lines = start + len(selected) < len(lines)

        body = "\n".join(
            f"{start + index + 1}: {line}" for index, line in enumerate(selected)
        )
        return {
            "path": self.workspace.relative_to_root(path),
            "content": body,
            "line_count": len(selected),
            "truncated": truncated_bytes or truncated_lines,
        }
    # ANCHOR_END: m2-read-tool

File: examples/rust/m2-tool-runtime/src/tools/read.rs
use crate::registry::{require_object, require_string};
use crate::tool_types::{Tool, ToolError, ToolSpec};
use crate::workspace::Workspace;
use serde_json::{Value, json};

const DEFAULT_MAX_LINES: usize = 2000;
const DEFAULT_MAX_BYTES: usize = 512_000;

/// 在 Workspace 内有界读取文本文件。
pub struct ReadTool {
    workspace: Workspace,
}

impl ReadTool {
    /// 绑定固定 Workspace。
    pub fn new(workspace: Workspace) -> Self {
        Self { workspace }
    }
}

impl Tool for ReadTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "read".into(),
            description: "Read file contents inside the workspace".into(),
            input_schema: json!({
                "type": "object",
                "required": ["path"],
                "properties": {
                    "path": {"type": "string"},
                    "offset": {"type": "integer"},
                    "limit": {"type": "integer"}
                },
                "additionalProperties": false
            }),
            strict: false,
        }
    }

    fn validate_arguments(&self, arguments: &Value) -> Result<(), ToolError> {
        let object = require_object(arguments)?;
        let _ = require_string(object, "path")?;
        if let Some(offset) = object.get("offset") {
            let value = offset
                .as_u64()
                .ok_or_else(|| ToolError::new("offset must be a positive integer"))?;
            if value < 1 {
                return Err(ToolError::new("offset must be a positive integer"));
            }
        }
        if let Some(limit) = object.get("limit") {
            let value = limit
                .as_u64()
                .ok_or_else(|| ToolError::new("limit must be a positive integer"))?;
            if value < 1 {
                return Err(ToolError::new("limit must be a positive integer"));
            }
        }
        Ok(())
    }

    fn execute(&mut self, arguments: &Value) -> Result<Value, ToolError> {
        let object = require_object(arguments)?;
        let path_arg = require_string(object, "path")?;
        let path = self.workspace.resolve(Some(path_arg), true)?;
        if !path.is_file() {
            return Err(ToolError::new(format!("not a file: {path_arg:?}")));
        }
        let mut data = std::fs::read(&path).map_err(|error| ToolError::new(error.to_string()))?;
        let mut truncated = false;
        if data.len() > DEFAULT_MAX_BYTES {
            data.truncate(DEFAULT_MAX_BYTES);
            truncated = true;
        }
        let text = String::from_utf8_lossy(&data);
        let lines: Vec<&str> = text.lines().collect();
        let offset = object
            .get("offset")
            .and_then(Value::as_u64)
            .unwrap_or(1)
            .max(1) as usize;
        let start = offset.saturating_sub(1);
        let limit = object
            .get("limit")
            .and_then(Value::as_u64)
            .map(|value| value as usize)
            .unwrap_or(DEFAULT_MAX_LINES);
        let end = (start + limit).min(lines.len());
        if end < lines.len() {
            truncated = true;
        }
        let body = lines[start..end]
            .iter()
            .enumerate()
            .map(|(index, line)| format!("{}: {line}", start + index + 1))
            .collect::<Vec<_>>()
            .join("\n");
        Ok(json!({
            "path": self.workspace.relative_to_root(&path)?,
            "content": body,
            "line_count": end.saturating_sub(start),
            "truncated": truncated,
        }))
    }
}
